### PR TITLE
 Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Innocent259**
In this project milestone we: 

1. Fixed n+1 problem for fetching all posts and their comment of a user
2. Used Capybara to write integration tests of views

Before solving the n+1 problem 

Screenshot from 2023-12-21 13-17-47

After solving the problem 

Screenshot from 2023-12-21 13-21-35

### CODE REVIEW

File: spec/features/users/show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      expect(page).to have_content("Comments: #{post.comments.count}")
+      expect(page).to have_content("Likes: #{post.likes.count}")
+    end
+  end
+  it 'displays a "See all posts" button if the user has more than 3 posts' do
+    expect(page).to have_selector('button', text: 'See all posts')
+  end
+  it 'redirects to the user posts index page when "See all posts" button is clicked' do
+    visit user_path(@user)
+    click_link 'See all posts'
```
Comment (Ibizugbe) [inline spec/features/users/show_spec.rb]: Nice job so far😉. Just as [required](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md) in the `user show page` kindly add the following tests😉.

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's first 3 posts.
- When I click a user's post, it redirects me to that post's show page.
image

File: spec/features/posts/index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  end
+
+  it 'displays some of the post body' do
+    expect(page).to have_content(@posts.first.text[0..149])
+  end
+
+  it 'displays the first comments on a post' do
+    expect(page).to have_content(@comments.first.text)
+  end
+
```
Comment (Ibizugbe) [inline spec/features/posts/index_spec.rb:62]: Nice job so far, Just as [required](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md), kindly do well to add the following tests😉.

- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page
image

File: spec/features/posts/show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    expect(page).to have_content(@post.text)
+    expect(page).to have_content('Comments: 5')
+    expect(page).to have_content('Likes: 0')
+  end
+
+  it 'displays comments for the post' do
+    @post.comments.each do |comment|
+      expect(page).to have_content(comment.user.name)
+      expect(page).to have_content(comment.text)
+    end
```
Comment (Ibizugbe) [inline spec/features/posts/show_spec.rb]: Nice job so far👍🏼. you seem to have missed some tests here😉. Just as [required](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md), kindly add the following tests👍🏼.

- I can see who wrote the post.
image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @Innocent259  @maths29 👋🏼👋🏼👋🏼, 

request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Linters are passing👍🏼.
- [x] `gitflow` is used👍🏼.
- [x] Tests for user index page has been written👍🏼.


## Required Changes ♻️

- [ ] In your PR, Kindly add a screenshot of your app's log before and after the fix and ensure all your tests are passing before proceeding to request a review😉. as this is a requirement😉.
image


_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/posts/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  it 'displays how many likes a post has' do
+    expect(page).to have_content("Likes: #{@posts.first.likes_counter}")
+  end
+
+  it 'displays a section for pagination if there are more posts than fit on the view' do
+    expect(page).to have_content('Pagination')
+  end
+end
```
Comment (OyePriscilla) [inline spec/features/posts/index_spec.rb]: - [ ] Please as requested earlier, kindly add this missing test that the previous reviewer highlighted. 👇 

Untitled

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @Innocent259 👍, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

so-close-so-close-seth-meyers

 ## Highlights
Solved n+1 queries problem ✔️ 
Implemented some test cases as highlighted in previous review ✔️ 

 ## Required Changes ♻️

- [ ] Kindly add correct screenshot of the n+1 queries problem solved; the screenshot for before and after running the queries are the same. 👍 

Untitled

Untitled

_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @Innocent259, and @maths29👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

image

## To Highlight 🎉
- All tests are passed. ✔️ 
- Good job fixing n+1 problem. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**